settings.txt
# This file specifies the key bindings for this game. The bindings are all
# numbers for pygame.K_KEYNAME, as listed at
# http://pygame.org/docs/ref/pygame_constants.html
up=273
down=274
left=276
right=275
action=13
cancel=27

# In battle, do a standard attack. (a)
attack=97
# In the main screen, save: (s)
save=115
# In the main screen, quit: (q)
quit=113
# In the main screen, open the inventory: (i)
inv=105

# If using a cheating character, load the console in the main window. (~)
load_console=96

# Difficulty: Easy=0, Normal=1, Hard=2.
difficulty=1

# Size (in pixels) of the editor screen
editor_xy_size=1024,768

# Is the game screen fullscreen, (or windowed)?
fullscreen=0


Changelog
# Changelog

Split into Active and Archive sections. Active is the current development version, and Archive is the preservation of the original author's changelog.

## Active

### Plans

    Get the game running(tm)
    Clean up the code of all commented out and unused code
    Transition everything to python3 conventions and best practices
    Hopefully document things a bit
    Scale up the resolution and assets slightly so that it is more playable on modern displays

### Versions

#### 4.0-dev

Currently in development so changlog is not complete or even accurate

    Complete conversion of all code to python3.12 and pygame 2.5x

## Archive

### Authors Plans

These were marked as TODO items in the changelog from the original author. I may or may not look into these, depending on how much time it takes just to get the game running again.

    Multiple Characters.
    map transitions.
    Add ability to "undo" in map editor.
    Complete testing map.
    Scripted Enemy attacks/ep.
    Allow new stats.
    Spawn monster.
    Scriptable equipment spaces.
    Show picture in popup window.
    Include files/functions.
    Better shop price adjustments.
    Actors.
    Better Shop scripting.
    Adjustment of dropped items.
    Number of used inv spots.
    Module in .zip.

## Historical Versions

These were mostly preserved as-is, with only minor formatting/spacing changes

### Version - 3.56

    Improved speed when loading linked maps.
    The save button in the inventory refreshes properly. (Lukas Sabota for report)
    Fixed a couple walkable mistakes in the first DA town. (Lukas Sabota for report)
    Joystick functionality added. (Lukas Sabota for suggestion)
    Psyco will be used if available.
    Added options screen.
    Loading screen works properly with 0 saves.
    Text entry no longer uses the custom keys.

### Version - 3.55

    Fixed bug that dropped an item when trying to use a skill. (report by YK)
    Game will no longer crash if no sound is available. (fix by Gekko04)
    The Death spell in Dark Ages actually works now. (report by Daniel Sturm)

### Version - 3.54

    Allowed skill usage outside of battle.
    Added copytile mode in editor.
    Added Dark Ages module.
    Manually setting exp levels is possible.

### Version - 3.53

    Added code to allow world-map type environments.
    Added fade-out effect. Now all I need is a fade-in effect...
    Improved the map editor significantly.
    Recreated the towns.

### Version - 3.52

    Rewrote the display code to use pygame instead of Tkinter. This should help in adding more graphical flash.
    Minor bugfixes.

### Version - 3.51

    Fixed a few bugs that ended up in 3.50.

### Version - 3.50

    UI re-organization
    unified game into a single canvas/window
    Completely removed PGUP/DN keys as unneeded.
    Fixed Python traceback when closing the main window with the inv open.
    skill("use", "something") added.
    skill use in battle shows price, items can be equipped during battle
    generic_dialog() added.
    inv("use", "something") added.
    mapspot() added.
    gamestat("new*") added.
    Stat names can be adjusted.
    New game/load game/battle/console/rpg.py/message scroller windows redone to use canvas exclusively.
    Dialogs work from battle.
    inv() can return item stats.
    equip() function added.
    All "clicking" binds use g.window_main, and <ButtonRelease-1>.
    Confirmation when closing game.

### Version - 3.49

    Redid inv; it now uses the same canvas as the main map.
    Redid store; it now uses the same canvas as the main map.
    Removed left sidebar, after transferring all data to other locations.
    Double-click in inv/store will affect item. (Michael Kallas for idea)
    Player will die properly when facing non-random enemies. (Michael Kallas for bug report)
    Added inv icon to main map.
    Better battle scripting (possible to adjust monster stats).
    Redid error checking slightly. Less chance for infinite loops now.
    Adjusted shop doors. (No moving two spaces on entry.)
    Added gamestat() command.
    Possible to have newlines in dialog/question.
    Fixed bug with hero().
    Added townsperson to island_shops.
    Script set() command accepts shortcuts.
    Can view xyz coords in scripting.
    addoverpix() scripting command added.
    inv() scripting command added.
    skill() scripting command added.
    inv_spot() added. (View item by inv loc.)
    Ability to change the titlebar text.
    Nerfed frenzy and dismember a bit.

### Version - 3.48

    Any images in modules/default/images will also be loaded. Completely useless right now, but if we get multiple modules, could reduce size.
    Skills can now be script-based.
    Map editor now correctly deals with multiple modules.
    Main window no longer comes up twice if "new game" is chosen twice rapidly.
    Added cheat console.
    Added wall_* script commands. (Allison Marles)
    Added/updated levels. (Allison Marles)
    Difficulty levels. (settings.txt)
    default_player_name actually works now.
    Updated docs.
    Shop name listed in the store, (Allison Marles)
    Prevent edge-of-map Python tracebacks.
    Redid scripting. All commands are now stackable.

### Version - 3.47

    Loadgame window displays stats before loading the game.
    Moved all player stats to the file "player.py", and changed it to a class.
    (This is the first step to getting multiple characters.)
    Added hurt_monster and damage_monster scripting commands.
    Cleaned up the battle code a bit.
    Minor improvements to a few maps.

### Version - 3.46

    Stopped map editor from deleting comments/indentation/blank lines.
    Fixed some minor bugs/CVS problems. (some fixes by Allison Marles)
    It is now possible to have items with scripting.
    Inventory equipment is tile-based.
    Stores, battles, and inventory all appear within the main window now.

### Version - 3.45

    Wingame.txt works properly again.

src/options.py
"""
This file controls the options screen.
"""

from os import path

import pygame

import g

# -1=none, 0=fullscreen, 1=difficulty, 2=custom key
# 5=save, 6=reset, 7=cancel
# 9=none, 10=Up, 11=Right, 12=Down, 13=Left, 14=Action, 15=Cancel, 16=reset,
# 17=back
global curr_button
curr_button = 1
global tmp_keys
tmp_keys = {}


# unfocused, inactive=gray; focused, inactive=green; focused, active=red
def box_with_text(xy, text, focused, active=False):
    in_color = "light_gray"
    if focused:
        in_color = "dh_green"
        if active:
            in_color = "light_red"
    g.create_norm_box((xy[0] + 2, xy[1] + 2), (176, 23), inner_color=in_color)
    g.print_string(g.screen, text, g.font, (xy[0] + 9, xy[1] + 10))


def refresh_window():
    # create the window
    x_start = config.TILESIZE * g.main.mapsizex / 2 - 90
    y_start = config.TILESIZE * g.main.mapsizey / 2 - 80
    x_width = 180
    g.create_norm_box((x_start, y_start), (x_width, 212))

    # fullscreen
    fullscreen_img = "check.png"
    if tmp_fullscreen == 1:
        fullscreen_img = "check_sel.png"
    in_color = "light_gray"
    if curr_button == 0:
        in_color = "dh_green"
    g.create_norm_box((x_start + 2, y_start + 2), (x_width - 4, 23), inner_color=in_color)
    g.screen.blit(g.buttons[fullscreen_img], (x_start + 7, y_start + 3))
    g.print_string(g.screen, "Fullscreen", g.font, (x_start + 27, y_start + 10))

    # difficulty
    if tmp_difficulty == 0:
        diff_string = "Difficulty: Easy"
    elif tmp_difficulty == 1:
        diff_string = "Difficulty: Normal"
    else:
        diff_string = "Difficulty: Hard"
    focused = False
    if curr_button == 1:
        focused = True
    box_with_text((x_start, y_start + 25), diff_string, focused)

    # custom keyboard
    focused = False
    if curr_button == 2:
        focused = True
    box_with_text((x_start, y_start + 50), "Customize Keyboard", focused)

    # save setttings
    focused = False
    if curr_button == 5:
        focused = True
    box_with_text((x_start, y_start + 135), "Apply and Save Settings", focused)

    # reset setttings
    focused = False
    if curr_button == 6:
        focused = True
    box_with_text((x_start, y_start + 160), "Reset Settings", focused)

    # leave
    focused = False
    if curr_button == 7:
        focused = True
    box_with_text((x_start, y_start + 185), "Cancel (Don't Save)", focused)

    pygame.display.flip()


def refresh_key_window(adjusting=False):
    # create the window
    x_start = config.TILESIZE * g.main.mapsizex / 2 - 90
    y_start = config.TILESIZE * g.main.mapsizey / 2 - 80
    x_width = 180
    g.create_norm_box((x_start, y_start), (x_width, 212))

    # up
    tmp_string = "Up: " + pygame.key.name(tmp_keys["up"])
    focused = False
    if curr_button == 10:
        focused = True
    box_with_text((x_start, y_start), tmp_string, focused, adjusting)

    # right
    tmp_string = "Right: " + pygame.key.name(tmp_keys["right"])
    focused = False
    if curr_button == 11:
        focused = True
    box_with_text((x_start, y_start + 25), tmp_string, focused, adjusting)

    # down
    tmp_string = "Down: " + pygame.key.name(tmp_keys["down"])
    focused = False
    if curr_button == 12:
        focused = True
    box_with_text((x_start, y_start + 50), tmp_string, focused, adjusting)

    # Left
    tmp_string = "Left: " + pygame.key.name(tmp_keys["left"])
    focused = False
    if curr_button == 13:
        focused = True
    box_with_text((x_start, y_start + 75), tmp_string, focused, adjusting)

    # Action
    tmp_string = "Action: " + pygame.key.name(tmp_keys["action"])
    focused = False
    if curr_button == 14:
        focused = True
    box_with_text((x_start, y_start + 100), tmp_string, focused, adjusting)

    # Cancel
    tmp_string = "Cancel: " + pygame.key.name(tmp_keys["cancel"])
    focused = False
    if curr_button == 15:
        focused = True
    box_with_text((x_start, y_start + 125), tmp_string, focused, adjusting)

    # reset setttings
    focused = False
    if curr_button == 16:
        focused = True
    box_with_text((x_start, y_start + 160), "Reset Settings", focused)

    # leave
    focused = False
    if curr_button == 17:
        focused = True
    box_with_text((x_start, y_start + 185), "Back", focused)

    pygame.display.flip()


# All keypresses pass through here. Based on the key name,
# give the right action. ("etc", "left", "right", "up", "down", "return")
def key_handler(switch):
    global curr_button
    global tmp_fullscreen
    global tmp_difficulty
    global tmp_keys
    if switch == g.bindings["cancel"]:
        cancel_settings()
    elif switch == g.bindings["up"]:
        curr_button -= 1
        if curr_button <= -1:
            curr_button = 7
    elif switch == g.bindings["down"]:
        curr_button += 1
        if curr_button > 7:
            curr_button = 0
    elif switch == g.bindings["action"]:
        if curr_button == 0:  # fullscreen
            if tmp_fullscreen == 0:
                tmp_fullscreen = 1
            else:
                tmp_fullscreen = 0
        elif curr_button == 1:  # difficulty
            if tmp_difficulty == 0:
                tmp_difficulty = 1
            elif tmp_difficulty == 1:
                tmp_difficulty = 2
            else:
                tmp_difficulty = 0
        elif curr_button == 2:  # custom key
            curr_button = 10
            custom_key()
            curr_button = 2
        elif curr_button == 5:  # save

            # apply settings to the game
            if g.fullscreen != tmp_fullscreen:
                pygame.display.toggle_fullscreen()
            g.fullscreen = tmp_fullscreen
            g.difficulty = tmp_difficulty
            g.bindings["up"] = tmp_keys["up"]
            g.bindings["down"] = tmp_keys["down"]
            g.bindings["right"] = tmp_keys["right"]
            g.bindings["left"] = tmp_keys["left"]
            g.bindings["action"] = tmp_keys["action"]
            g.bindings["cancel"] = tmp_keys["cancel"]
            # Restore any shortcuts that had been removed before.
            g.bindings["attack"] = pygame.K_a
            g.bindings["save"] = pygame.K_s
            g.bindings["quit"] = pygame.K_q
            g.bindings["inv"] = pygame.K_i
            g.bindings["load_console"] = pygame.K_BACKQUOTE
            # Remove any shortcuts that are already taken.
            for binding in ["up", "down", "right", "left", "action", "cancel"]:
                if g.bindings[binding] == pygame.K_a:
                    g.bindings["attack"] = 0
                if g.bindings[binding] == pygame.K_s:
                    g.bindings["save"] = 0
                if g.bindings[binding] == pygame.K_q:
                    g.bindings["quit"] = 0
                if g.bindings[binding] == pygame.K_i:
                    g.bindings["inv"] = 0
                if g.bindings[binding] == pygame.K_BACKQUOTE:
                    g.bindings["load_console"] = 0

            # save the settings to disk
            options_file_text = []
            if path.exists(g.mod_directory + "/../../settings.txt") == 1:
                optionfile = open(g.mod_directory + "/../../settings.txt")
                options_file_text = optionfile.readlines()
            else:
                "Could not open settings.txt. Not saving settings."
                cancel_settings()
                refresh_window()
                return

            for linenum in range(len(options_file_text)):
                command = options_file_text[linenum].split("=")[0].strip().lower()
                if (
                    command == "up"
                    or command == "down"
                    or command == "left"
                    or command == "right"
                    or command == "action"
                    or command == "cancel"
                    or command == "attack"
                    or command == "save"
                    or command == "quit"
                    or command == "inv"
                    or command == "load_console"
                ):
                    options_file_text[linenum] = command + "=" + str(g.bindings[command]) + "\n"
                if command == "difficulty":
                    options_file_text[linenum] = command + "=" + str(g.difficulty) + "\n"
                if command == "fullscreen":
                    options_file_text[linenum] = command + "=" + str(g.fullscreen) + "\n"

            optionfile.close()
            optionfile = open(g.mod_directory + "/../../settings.txt", "w")
            optionfile.writelines(options_file_text)

            cancel_settings()
        elif curr_button == 6:  # reset
            tmp_fullscreen = 0
            tmp_difficulty = 1
            tmp_keys = {}
            tmp_keys["up"] = pygame.K_UP
            tmp_keys["down"] = pygame.K_DOWN
            tmp_keys["right"] = pygame.K_RIGHT
            tmp_keys["left"] = pygame.K_LEFT
            tmp_keys["action"] = pygame.K_RETURN
            tmp_keys["cancel"] = pygame.K_ESCAPE
        elif curr_button == 7:
            cancel_settings()  # cancel
    refresh_window()


def bind_new_key(start_key):
    while True:
        pygame.time.wait(30)
        g.clock.tick(30)
        if g.break_one_loop > 0:
            g.break_one_loop -= 1
            break
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return
            elif event.type == pygame.KEYDOWN:
                return event.key
    return start_key


# key handler for the key customizer window.
def key_key_handler(switch):
    global curr_button
    global tmp_keys
    if switch == g.bindings["cancel"]:
        cancel_settings()
    elif switch == g.bindings["up"]:
        curr_button -= 1
        if curr_button <= 9:
            curr_button = 17
    elif switch == g.bindings["down"]:
        curr_button += 1
        if curr_button > 17:
            curr_button = 10
    elif switch == g.bindings["action"]:
        if curr_button == 10:  # up
            refresh_key_window(True)
            tmp_keys["up"] = bind_new_key(tmp_keys["up"])
        if curr_button == 11:  # right
            refresh_key_window(True)
            tmp_keys["right"] = bind_new_key(tmp_keys["right"])
        if curr_button == 12:  # down
            refresh_key_window(True)
            tmp_keys["down"] = bind_new_key(tmp_keys["down"])
        if curr_button == 13:  # left
            refresh_key_window(True)
            tmp_keys["left"] = bind_new_key(tmp_keys["left"])
        if curr_button == 14:  # action
            refresh_key_window(True)
            tmp_keys["action"] = bind_new_key(tmp_keys["action"])
        if curr_button == 15:  # cancel
            refresh_key_window(True)
            tmp_keys["cancel"] = bind_new_key(tmp_keys["cancel"])
        if curr_button == 16:  # reset
            tmp_keys = {}
            tmp_keys["up"] = g.bindings["up"]
            tmp_keys["down"] = g.bindings["down"]
            tmp_keys["right"] = g.bindings["right"]
            tmp_keys["left"] = g.bindings["left"]
            tmp_keys["action"] = g.bindings["action"]
            tmp_keys["cancel"] = g.bindings["cancel"]
        if curr_button == 17:  # back
            cancel_settings()
    refresh_key_window()


def custom_key():
    refresh_key_window()
    while True:
        pygame.time.wait(30)
        g.clock.tick(30)
        if g.break_one_loop > 0:
            g.break_one_loop -= 1
            break
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return
            elif event.type == pygame.KEYDOWN:
                key_key_handler(event.key)
                repeat_key = 0
            elif event.type == pygame.MOUSEMOTION:
                mouse_handler_move(event.pos)
            elif event.type == pygame.MOUSEBUTTONDOWN:
                if mouse_handler_down(event.pos) == 1:
                    return


def mouse_handler_move(pos):
    pos = (pos[0], pos[1] - config.TILESIZE * g.main.mapsizey / 2 + 80)
    global curr_button
    if (
        pos[0] < config.TILESIZE * g.main.mapsizex / 2 - 90
        or pos[0] > config.TILESIZE * g.main.mapsizex / 2 + 90
        or pos[1] < 0
        or pos[1] > 212
    ):
        if curr_button >= 19:
            curr_button = 19
        elif curr_button >= 9:
            curr_button = 9
            refresh_key_window()
        else:
            curr_button = -1
            refresh_window()
    elif curr_button >= 19:
        if pos[1] < 25:
            curr_button = 20
        elif pos[1] < 50:
            curr_button = 21
        elif pos[1] < 75:
            curr_button = 22
        elif pos[1] < 100:
            curr_button = 23
        elif pos[1] < 160:
            curr_button = 19
        elif pos[1] < 185:
            curr_button = 24
        elif pos[1] < 210:
            curr_button = 25
    elif curr_button >= 9:
        if pos[1] < 25:
            curr_button = 10
        elif pos[1] < 50:
            curr_button = 11
        elif pos[1] < 75:
            curr_button = 12
        elif pos[1] < 100:
            curr_button = 13
        elif pos[1] < 125:
            curr_button = 14
        elif pos[1] < 150:
            curr_button = 15
        elif pos[1] < 160:
            curr_button = 9
        elif pos[1] < 185:
            curr_button = 16
        elif pos[1] < 210:
            curr_button = 17
        refresh_key_window()
    else:
        if pos[1] < 25:
            curr_button = 0
        elif pos[1] < 50:
            curr_button = 1
        elif pos[1] < 75:
            curr_button = 2
        elif pos[1] < 100:
            curr_button = 3
        elif pos[1] < 125:
            curr_button = 4
        elif pos[1] < 135:
            curr_button = -1
        elif pos[1] < 160:
            curr_button = 5
        elif pos[1] < 185:
            curr_button = 6
        elif pos[1] < 210:
            curr_button = 7
        refresh_window()


def mouse_handler_down(pos):
    global curr_button
    if curr_button >= 9:
        key_key_handler(pygame.K_RETURN)
    else:
        key_handler(pygame.K_RETURN)


def init_window_options():
    global curr_button
    curr_button = 0

    global tmp_fullscreen
    global tmp_difficulty
    global tmp_keys
    tmp_fullscreen = g.fullscreen
    tmp_difficulty = g.difficulty
    tmp_keys = {}
    tmp_keys["up"] = g.bindings["up"]
    tmp_keys["down"] = g.bindings["down"]
    tmp_keys["right"] = g.bindings["right"]
    tmp_keys["left"] = g.bindings["left"]
    tmp_keys["action"] = g.bindings["action"]
    tmp_keys["cancel"] = g.bindings["cancel"]
    refresh_window()
    while True:
        pygame.time.wait(30)
        g.clock.tick(30)
        if g.break_one_loop > 0:
            g.break_one_loop -= 1
            break
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return
            elif event.type == pygame.KEYDOWN:
                key_handler(event.key)
                repeat_key = 0
            elif event.type == pygame.MOUSEMOTION:
                mouse_handler_move(event.pos)
            elif event.type == pygame.MOUSEBUTTONDOWN:
                if mouse_handler_down(event.pos) == 1:
                    return


def cancel_settings():
    g.break_one_loop = 1
